### ISSUE TITLE
Two Tower Model -Incremental Training

### ISSUE THREAD

**Issue by nicewenhui** [caesarjuly/reginx#4]
Hi, I'm new to recommender systerms , and your codes and posts helped me a lot.  

I have a stupid question I want to ask. In the data preparation step, I found that you used StringLookup to generate vocabulary for user_id, and then generated embeddings using embedding layers. In the real world, new users and new content continue to emerge, but I do not want to retrain the whole model. I would like to use incremental training. In this case, I don't know how to handle the new users and items. I found that you used the MutableHashTable to clear expired old keys. Do you think this approach is also appropriate for my case? I have tried using tf.keras.layers.Hashing, but I found the hash collection to be quite large.

**Comment by caesarjuly** [comment]
Hash trick is a common practice when dealing with huge sparse id features like userIds. It can achieve a relatively good performance with limited memory consumption. 
![image](https://github.com/caesarjuly/reginx/assets/927521/ee873c6f-fdcb-4497-ab71-78ddfd270c22)

If using MutableHashTable, it becomes a dynamic growing userId lookup table which shares the similar idea as the [dynamic embedding layer](https://blog.tensorflow.org/2023/04/training-recommendation-model-with-dynamic-embeddings.html). The only concern is the unlimited memory consumption. A common approach to solve this is to use a distributed database to store all the embedding data as what the dynamic embedding layer did. 
In summary:

1. if your userId cardinality is small which means limited hash collision, there will be little difference between using Hashing and MutableHashTable.
2. ~~if your userId cardinality is medium and can be stored in one single machine, MutableHashTable should be a good choice.~~
3. if your userId cardinality is huge, dynamic embedding layer is what you need

**Comment by nicewenhui** [comment]
Thanks for your detailed explanation, it really helps me a lot!

**Comment by nicewenhui** [comment]
Hi,

I still have some unclear problems.

I found that when using the MutableHashTable, it creates a key-embedding table that is used for quick lookup and sets a default value, such as 0, for unseen keys.

However, I'm a bit confused about how this works for incremental training. For example, I've trained a base model, and the user_id_embedding table saved 10,000 unique user_id values and their corresponding user_id_embeddings. I also saved the trained model. In my next training session, I upload the trained model and use the previously trained parameters for incremental training. For instance, this time I only have 5,000 unique users, 1,000 of whom are new user_id values.

In this case, how will the embedding layer deal with the new user_id values? Will all the new user_id values be set to the default 0 embedding? What I want is for the new user_id values to generate their new embeddings and be inserted into the key-embedding table. Do you think this is possible?

Can you provide a code snapshot for this? I really don't know how to proceed. Thanks a lot.

**Comment by caesarjuly** [comment]
> Hi,
> 
> I still have some unclear problems.
> 
> I found that when using the MutableHashTable, it creates a key-embedding table that is used for quick lookup and sets a default value, such as 0, for unseen keys.
> 
> However, I'm a bit confused about how this works for incremental training. For example, I've trained a base model, and the user_id_embedding table saved 10,000 unique user_id values and their corresponding user_id_embeddings. I also saved the trained model. In my next training session, I upload the trained model and use the previously trained parameters for incremental training. For instance, this time I only have 5,000 unique users, 1,000 of whom are new user_id values.
> 
> In this case, how will the embedding layer deal with the new user_id values? Will all the new user_id values be set to the default 0 embedding? What I want is for the new user_id values to generate their new embeddings and be inserted into the key-embedding table. Do you think this is possible?
> 
> Can you provide a code snapshot for this? I really don't know how to proceed. Thanks a lot.

Ah, I made a mistake. It's been a while since I wrote model code. Now I fully understand the issue here:
1. TensorFlow embedding layer is not only a hash table that stores the key, embedding value mapping. It also support gradient back propagation for updating the embedding value. However, the MutableHashTable is only a hash table and cannot be updated dynamically. This means it's not a embedding layer, so it cannot be used as a embedding table.
2. So the only options we have is either using a hash layer to map userIds into the index of a fixed size embedding table . The structure is like userId -> hash layer -> embedding index -> embedding layer -> embedding value. The other option is to leverage the dynamic embedding layer I mentioned previously.
3. In my previous company, there is a third option, that is modifying the TensorFlow source code and customize the embedding layer to allow dynamic growth of the embedding table. I don't recommend this approach since it takes a lot of effort and hard to maintain.
